### ISSUE TITLE
Compiler is not found

### ISSUE THREAD

**Issue by scheibelp** [LLNL/spack#3434]
Issue was originally reported by @BarrySmith in https://github.com/LLNL/spack/pull/3427 (copied here):

it does not even detect the clang compilers period! I used git bisect and found

e727f56d89ccd2aa9aabc14cf4efb471f32341dc is the first bad commit

In other words before this bad commit #2999 I would get

```
==> Added 3 new compilers to /Users/barrysmith/.spack/darwin/compilers.yaml
    gcc@6.3.0  gcc@4.2.1  clang@8.0.0-apple
```

Afterwards I get

```
==> Added 3 new compilers to /Users/barrysmith/.spack/darwin/compilers.yaml
    gcc@6.3.0  gcc@4.2.1
```

The "bad" commit it is pretty big so I don't know where to start in determining how the commit made clang unfindable.

**Comment by scheibelp** [comment]
@BarrySmith apologies but I realized this should be opened as a separate issue so I copied my questions to you into this thread

I want to be 100% sure, does your output in fact look exactly like:

```
==> Added 3 new compilers to /Users/barrysmith/.spack/darwin/compilers.yaml
    gcc@6.3.0  gcc@4.2.1
```

? I would have expected it to look like:

```
==> Added 2 new compilers to /Users/barrysmith/.spack/darwin/compilers.yaml
    gcc@6.3.0  gcc@4.2.1
```

(only difference is "Added 2 new compilers" vs. "Added 3 new compilers")

Also: is there an entry for clang in ```/Users/barrysmith/.spack/compilers.yaml``` (which omits the ```darwin``` prefix)? If so, can you try manually removing it and check whether ```spack compiler find``` finds clang for you?

Thanks

**Comment by scheibelp** [comment]
This is related to https://github.com/LLNL/spack/issues/3428 but in that case the compilers are actually found (but a confusing error message says otherwise). In this case the compiler is actually not added to the configuration.

**Comment by BarrySmith** [comment]
Sorry, my mistake, I was lazy when I cut and pasted the message and just removed the Clang from the list. It does indicate only 2. 

```
Added 3 new compilers to /Users/barrysmith/.spack/darwin/compilers.yaml
    gcc@6.3.0  gcc@4.2.1
```

I manully removed the .spack/darwin/compilers.yaml before each run and then checked the file afterwards and it listed the same items as was printed to stdout. So it truly is not finding the clang.

**Comment by scheibelp** [comment]
> So it truly is not finding the clang.

To make sure of that can you see whether ```spack compiler list``` outputs clang?

Part of what I added in #2999 was to avoid re-adding an entry if it already existed, there's a few potential config files other than the one you manually removed (and if clang shows up in the listing it would be in one of those):

* ~/.spack/darwin/compilers.yaml
* ~/.spack/compilers.yaml
* [spack git repo]/etc/spack/site/compilers.yaml
* [spack git repo]/etc/spack/site/darwin/compilers.yaml

It sounds like you have manually removed the first. I think the second may exist on your system. I can improve this so it isn't as confusing in the future.

**Comment by BarrySmith** [comment]
Crap, my fault. Sorry I wasted your time. I deleted the ~/.spack/darwin/compilers.yaml but didn't think about the ~/.spack/compilers.yaml file (which only had clang listed, which is why I got the messages only about finding the other compilers, since it alreay knew about clang). When I removed both files spack did detect all the compilers.

Having both locations ~/.spack/darwin/compilers.yaml and ~/.spack/compilers.yaml is problematic! If you want to support both maybe when both files exist the spack compiler find (and list) should print both filenames to the screen so at least the user is aware of both. Also when find finds nothing new it should still print the names of the files it found so people are aware of them. Personally I woud remove support for ~/.spack/compilers.yaml from spack; its going to cause more bad bug reports like mine. @tgamblin
